### PR TITLE
Add support for default value in macros

### LINKED ISSUE #198: Add support for macro argument default value
It's possible in [jinja](https://jinja.palletsprojects.com/en/3.1.x/templates/#macros) to have default value for macro arguments:

```jinja
{% macro input(name, value='', type='text', size=20) -%}
    <input type="{{ type }}" name="{{ name }}" value="{{value|e }}" size="{{ size }}">
{%- endmacro %}
```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #198.

So I went the easy way and allowed to have an expression as default value. Not sure if it's a good or bad idea.

I still need to update the book, but before doing that, I want to ensure with you if expression as default value is ok with you or if I should limit it.

### CODE REVIEW

**Comment by Kijewski** [comment]
> So I went the easy way and allowed to have an expression as default value. Not sure if it's a good or bad idea.

I think that's a good idea! The only question is when to evaluate the expressions:

```jinja
{% macro sum(a=0, b=a, c=a+b, d=a+b+c) %}
  {{a}} {{b}} {{c}} {{d}}
{% endmacro %}
```

In a call to `sum`, should `a` refer to the the parameter `a`, or to an identifier `a` at the calling location? IMO, the former would be vastly more useful.

**Comment by GuillaumeGomez** [comment]
Currently it's to the parameter. But gonna add a test for that.

Well then, since you like it as is, I'll continue on this implementation. :)

**Comment by GuillaumeGomez** [comment]
Ready for review. :)

File: rinja_parser/src/node.rs [hunk 1]
```diff
@@ -568,17 +569,41 @@ impl<'a> Loop<'a> {
...
+macro_rules! check_duplicated_name {
+    ($names:ident, $arg_name:ident, $i:ident) => {
+        if !$names.insert($arg_name) {
+            return Err(nom::Err::Failure(ErrorContext::new(
+                format!("duplicated argument `{}`", $arg_name),
+                $i,
+            )));
+        }
+    };
+}
```
Comment (Kijewski) [inline rinja_parser/src/node.rs:586]: Why didn't you write that as a function?

File: rinja_derive/src/generator.rs [hunk 2]
```diff
@@ -2959,11 +2978,12 @@ fn is_copyable(expr: &Expr<'_>) -> bool {
...
-        Expr::BoolLit(_) | Expr::NumLit(_, _) | Expr::StrLit(_) | Expr::CharLit(_) => true,
+        Expr::BoolLit(_)
+        | Expr::NumLit(_, _)
+        | Expr::StrLit(_)
+        | Expr::CharLit(_)
+        | Expr::BinOp(_, _, _) => true,
         Expr::Unary(.., expr) => is_copyable_within_op(expr, true),
-        Expr::BinOp(_, lhs, rhs) => {
-            is_copyable_within_op(lhs, true) && is_copyable_within_op(rhs, true)
-        }
```
Comment (Kijewski) [inline rinja_derive/src/generator.rs:2966]: Is this change correct?

File: rinja_parser/src/node.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply rinja_parser/src/node.rs:586]: I did originally, but forgot why it didn't work out. Gonna give it another try.

File: rinja_derive/src/generator.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply rinja_derive/src/generator.rs:2966]: Yes. I discovered that if you do `a + b`, it was generated as `&(a + b)`, which is wrong. I encountered this bug when allowing expressions with macro params as default value.

**Comment by GuillaumeGomez** [comment]
Converted `check_duplicated_name` into a function again.

**Comment by GuillaumeGomez** [comment]
@Kijewski Forgotten or scared to look at it? 😆

**Comment by Kijewski** [comment]
Scared to look at it. :D But I'll to it tomorrow!

**Comment by GuillaumeGomez** [comment]
Only if you have time and motivation. 😉 And thanks in advance!

**Comment by Kijewski** [comment]
I refactored the code a bit to make it more easily understandable to me, and improved the error message for broken argument lists a bit.

Sorry for taking so long to look into this PR! Thank you for implementing it!

**Review by Kijewski — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks for the improvements (really loved the commit message "kill the Cows" 🤣 )!

**Pull request merged**